### PR TITLE
Initial version of complex layout performance test

### PR DESCRIPTION
**Pull request by vlidholt**

### CODE REVIEW

File: dev/benchmarks/complex_layout/lib/drawer.dart [hunk 1]
```diff
@@ -0,0 +1,91 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart' show timeDilation;
+import 'package:flutter/widgets.dart';
+
+import 'main.dart';
```
Comment (abarth) [inline dev/benchmarks/complex_layout/lib/drawer.dart]: Given that these two files import each other, maybe we should just combine all the code into one file.

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+  bool _lightTheme = true;
+  bool get lightTheme => _lightTheme;
+  void set lightTheme(bool value) {
+    setState(() {
+      _lightTheme = value;
+    });
+  }
+}
+
+class FlutterDemo extends StatefulWidget {
```
Comment (abarth) [inline dev/benchmarks/complex_layout/lib/main.dart]: Maybe rename these classes from "flutter demo" to something about "complex layout" ?

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+}
+
+class ItemGalleryBox extends StatelessWidget {
+  ItemGalleryBox(this.index);
+
+  final int index;
+
+  @override
+  Widget build(BuildContext context) {
+    List<String>tabNames = <String>[
```
Comment (Hixie) [inline dev/benchmarks/complex_layout/lib/main.dart]: nit: space missing before identifier name

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+      body: new Column(
+        children: <Widget>[
+          new Flexible(
+            child: new ScrollableMixedWidgetList(
+              key: const ValueKey("main-scroll"),
+              builder: (BuildContext context, int index) {
+                if (index % 2 == 0)
+                  return new FancyImageItem(index, key: new ValueKey("Item $index"));
+                else
+                  return new FancyGalleryItem(index, key: new ValueKey("Item $index"));
```
Comment (abarth) [inline dev/benchmarks/complex_layout/lib/main.dart]: You can just say `new Key`.  It defaults to making a `ValueKey<String>`.

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+    List<String>tabNames = <String>[
+      "A", "B", "C", "D"
+    ];
+
+    return new SizedBox(
+      height: 200.0,
+      child: new TabBarSelection<String>(
+        values: tabNames,
+        child: new Builder(
+          builder: (BuildContext context) {
```
Comment (Hixie) [inline dev/benchmarks/complex_layout/lib/main.dart]: why the builder?

**Comment by abarth** [comment]
LGTM

There's kind of a lot of boilerplate in here around getting all the Android and iOS manifests.  Maybe that's not worth worrying about but as we add more of these benchmarks, we might want to combine them into a single app to share on boilerplate.

File: dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        // Scroll up
+        for (int i = 0; i < 5; i++) {
+          await driver.scroll(stockList, 0.0, 300.0, new Duration(milliseconds: 300));
+          await new Future<Null>.delayed(new Duration(milliseconds: 500));
+        }
+      });
+
+      // Usually the profile is saved to a file and then analyzed using
+      // chrom://tracing or a script. Both are out of scope for this little
+      // test, so all we do is check that we received something.
```
Comment (Hixie) [inline dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart]: this comment contradicts the fact that you actually do save it lower down, no?

**Comment by Hixie** [comment]
Did you verify that the benchmark is affected by making the code slower?

**Comment by Hixie** [comment]
LGTM

File: dev/benchmarks/complex_layout/lib/drawer.dart (hunk 1 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/drawer.dart]: Fixing

File: dev/benchmarks/complex_layout/lib/main.dart (hunk 2 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/main.dart]: Fixing

File: dev/benchmarks/complex_layout/lib/main.dart (hunk 3 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/main.dart]: Fixing

File: dev/benchmarks/complex_layout/lib/main.dart (hunk 4 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/main.dart]: Fixing (including other places)

File: dev/benchmarks/complex_layout/lib/main.dart (hunk 5 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/main.dart]: Copy pasted the code from the examples, will see if it works without the builder too

File: dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart (hunk 6 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/test_driver/scroll_perf_test.dart]: Removed comment

**Comment by vlidholt** [comment]
> Did you verify that the benchmark is affected by making the code slower?

It currently just outputs a tracing file. Yegor is working on outputting summarized data, may hold off the verification until then.

File: dev/benchmarks/complex_layout/lib/main.dart (hunk 5 above)
Comment (vlidholt) [inline dev/benchmarks/complex_layout/lib/main.dart]: Totally works without the builder, this should probably be removed from the examples too.

**Pull request merged**